### PR TITLE
Enumeratees with different parsing modes

### PR DESCRIPTION
**Pull request by ImLiar**
Related to: https://github.com/circe/circe/issues/711

Additional enumeratees for `circe-streaming` with different `AsyncParser.Mode`s

### CODE REVIEW

File: modules/streaming/src/main/scala/io/circe/streaming/ParsingConfiguration.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package io.circe.streaming
+
+import jawn.AsyncParser
+
+case class ParsingConfiguration(parseMode: AsyncParser.Mode)
+
+object ParsingConfiguration {
+
+  implicit val defaultParsingConfiguration: ParsingConfiguration =
```
Comment (vkostyukov) [inline modules/streaming/src/main/scala/io/circe/streaming/ParsingConfiguration.scala]: I'm not a huge fan of implicit parameters used for configuration. I'd rather prefer us to be explicit about this. One way of doing this is to encode the parsing mode in the function name (e.g., `stringStreamParser` and `byteStreamParser` - maybe there are better names).

An alternative solution would be to add a new parsing mode to Jawn (something like `Smart` mode) when it tries to use `StreamValue` by default and then fall back to `UnwrapArray` if the top-level object is an array.

Comment (ImLiar) [inline modules/streaming/src/main/scala/io/circe/streaming/ParsingConfiguration.scala]: @vkostyukov I'm afraid it's not possible to distinguish `SingleValue` and `StreamValue` modes easily

Comment (ImLiar) [inline modules/streaming/src/main/scala/io/circe/streaming/ParsingConfiguration.scala]: @vkostyukov I'd go with implicitis for the sake of backward compatibility. In other case it'd be required to rename old functions in favour of new logic.

Comment (vkostyukov) [inline modules/streaming/src/main/scala/io/circe/streaming/ParsingConfiguration.scala]: I don't think we have to rename. We can keep `stringParser`/`byteParser` in place and just document what parsing mode they use.

I guess I'm missing something, but I'm not sure I understand the reason we need `SingleValue` given `StreamValue` exists.

**Comment by travisbrown** [comment]
I'm traveling right now but I'm also -1 on implicit configuration, and would much prefer breaking compatibility than adding it here.

File: modules/streaming/src/main/scala/io/circe/streaming/package.scala [hunk 2]
```diff
@@ -6,23 +6,58 @@ import io.circe.jawn.CirceSupportParser
...
+  /**
+    * String parser of JSON values wrapped in a single array
+    */
+  final def stringArrayParser[F[_]](implicit F: ApplicativeError[F, Throwable]): Enumeratee[F, String, Json] =
+    stringParser(AsyncParser.UnwrapArray)
+
+  /**
+    * String parser of JSON values delimited by whitespace/newline
+    */
+  final def stringStreamParser[F[_]](implicit F: ApplicativeError[F, Throwable]): Enumeratee[F, String, Json] =
```
Comment (vkostyukov) [inline modules/streaming/src/main/scala/io/circe/streaming/package.scala]: I like it. Even though I can totally see how stream parser vs. array parser may be misinterpreted as the "parser that takes stream or array as an input", I think these names our best bet (given we document the difference).

**Comment by ImLiar** [comment]
👍

**Comment by travisbrown** [comment]
At a first pass this looks good to me—I'll aim to get it merged later today.

**Comment by vkostyukov** [comment]
@travisbrown @ImLiar Anything's left here?

**Comment by travisbrown** [comment]
No, just me getting my act together. :smile:

By the way I was hoping to publish an iteratee.io 0.13.0 release this week, before the circe 0.9 milestone. The main thing I'm doing in the 0.13 release is using the new cats-effect type classes in iteratee-files, and providing a module for `cats.effect.IO`. (I hoped to have this done last weekend but I'm moving across the country in a couple of weeks and all my weekends have been totally lost to moving- and packing-related tasks recently.)

**Comment by vkostyukov** [comment]
Thanks, Travis! The moving burden is totally understandable - even moving within a city makes me crazy nervous!

I think I will cut a 0.16-M2 Finch release once iteratee is available.

**Comment by ImLiar** [comment]
Oh, this is awesome! Thanks
Waiting for new release 👍

**Pull request merged**